Update trimming doc (#36487)
dotnet/docs: docs/core/deploying/trimming/prepare-libraries-for-trimming.md

* Describes how to prepare libraries for trimming.
* Provides recommendations for [resolving common trimming warnings](#resolve-trim-warnings).

## Prerequisites

[.NET 6 SDK](https://dotnet.microsoft.com/download/dotnet) or later.

## Enable library trim warnings

Trim warnings in a library can be found with either of the following methods:

* Enabling project-specific trimming using the `IsTrimmable` property.
* Creating a trimming test app that uses all the APIs in the library, and enabling trimming for the test app.

We recommend using both approaches. Project-specific trimming is convenient and shows trim warnings for one project, but relies on the references being marked trim-compatible to see all warnings. Trimming a test app is more work, but shows all warnings.

### Enable project-specific trimming

### [.NET 6](#tab/net6)

To get the latest version of the analyzer with the most coverage, install the [.NET 8 SDK](https://dotnet.microsoft.com/download/dotnet) or later. Installing the .NET 8 SDK or later:

> * Updates the tooling used to build an app or library and enable trim warnings.

Reviewer: The recommendation should be for the trimming test app to target latest TFM for sure. And consider multi-targeting for the library itself where one of the TFMs is 6+ (or ideally 8+).
Sven - where's the doc about the recommended TFMs and the whole `IsTrimmable` story we did for 8?

> * Updates the tooling used to build an app or library and enable trim warnings.

Reviewer: Additionally it's confusing as written - .NET 6 tab recommends installing .NET 8 SDK, but .NET 7 tab doesn't say anything, so it seems .NET 7 SDK is fine which looks wrong.

In the following example, an unknown [Type](/dotnet/api/system.type) flows into the annotated method parameter. The unknown `Type` is from a field:

:::code language="csharp" source="~/docs/core/deploying/trimming/snippets/MyLibrary/Class1.cs" id="snippet_UMH":::

Similarly, here the problem is that the field `type` is passed into a parameter with these requirements. It's fixed by adding  [`[DynamicallyAccessedMembers]`](xref:System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute) to the field. `[DynamicallyAccessedMembers]` warns about code that assigns incompatible values to the field. Sometimes this process continues until a public API is annotated, and other times it ends when a concrete type flows into a location with these requirements. For example:

:::code language="csharp" source="~/docs/core/deploying/trimming/snippets/MyLibrary/Class1.cs" id="snippet_UMH2" highlight="1":::

In this case, the trim analysis keeps public methods of <xref:System.Tuple>, and produces further warnings.

## Recommendations

***Avoid*** reflection when possible.

* When using reflection, minimize reflection scope so that it's reachable only from a small part of the library.

> * Avoid using <!-- Review change: non-understood--> non-trimmable patterns in places like static constructors. Static constructors result in the warning propagating to all members of the class.

Author: Review

* Avoid annotating virtual methods or interface methods. Annotating virtual or interface methods requires all overrides to have matching annotations.

> * In some cases, you're able to mechanically propagate warnings through the library code without issues. <!--Review: What does mechanically propagate mean? For sure it won't MT (Machine Translate) --> Sometimes this results in much of your public API being annotated with [`[RequiresUnreferencedCode]`](xref:System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute). Annotating with `[RequiresUnreferencedCode]` is the right thing to do if the library behaves in ways that can't be understood statically by the trim analysis.

Author: manually propagate? As in add attributes in the call chain?

> # Prepare .NET libraries for trimming

Reviewer: As a general comment - we've made some improvements in this are for .NET 8. And should update this doc to include .NET 8 - and recommend using it.

There's another consideration which is mentioned already but it might be worth pointing it out more prominently. Higher version SDK can build libraries targeting lower version TFMs and the higher version SDK will have fixes and improvements. So we should pretty much always recommend using the latest SDK available (ideally .NET 8 already). This has very direct impact on the trimming story - the examples here for the project files will look somewhat different (and hopefully simpler) for .NET 8.

Finally, recently we've learned that multi-targeting is not only very common, but basically a must have when enabling trimming. (This is based on experience trying to make some NuGets trim compatible). I think the document should have a section about the multi-targeting and our recommendations around it. But the core point is that very few libraries will only target latest TFM. By far the most common case is either library targets `netstandard2.0` or it multi-targets `net472`, `netstandard2.0` and `net6.0` (or some combination of these).

I'm writing all this down so that we don't forget - it may well need to be done in multiple PRs and updates to the docs. But to have a good consistent story for library trimming (and even more importantly for AOT compatibility), we will need to include .NET8 and the above topics.

Author: > Finally, recently we've learned that multi-targeting is not only very common, but basically a must have when enabling trimming. (This is based on experience trying to make some NuGets trim compatible). I think the document should have a section about the multi-targeting and our recommendations around it. But the core point is that very few libraries will only target latest TFM. By far the most common case is either library targets `netstandard2.0` or it multi-targets `net472`, `netstandard2.0` and `net6.0` (or some combination of these).

@vitek-karas  see #36730 
If you can get someone to draft it, I'll clean it up.

The .NET SDK makes it possible to reduce the size of self-contained apps by [trimming](trim-self-contained.md). Trimming removes unused code from the app and its dependencies. Not all code is compatible with trimming. .NET provides trim analysis warnings to detect patterns that may break trimmed apps. This article:

* Describes how to prepare libraries for trimming.
* Provides recommendations for [resolving common trimming warnings](#resolve-trim-warnings).

## Prerequisites

[.NET 6 SDK](https://dotnet.microsoft.com/download/dotnet) or later.

## Enable library trim warnings

Trim warnings in a library can be found with either of the following methods:

* Enabling project-specific trimming using the `IsTrimmable` property.

> * Creating a trimming test app that uses all the APIs in the library, and enabling trimming for the test app.

Reviewer: This is not precise - while it's technically true that the test app will "reference" all of the APIs of the library, it is not done manually - the sentence here sounds like the developer will have to write an app which exercises all APIs - which sounds very tedious.

Setting `<IsTrimmable>true</IsTrimmable>` marks the assembly as "trimmable" and enables trim warnings. "trimmable" means the project:

* Is considered compatible with trimming.
* Shouldn't generate trim related warnings when building. When used in a trimmed app, the assembly has its unused members trimmed in the final output.

The `IsTrimmable` property defaults to `true` when configuring a project as AOT-compatible with `<IsAotCompatible>true</IsAotCompatible>`. For more information, see [AOT-compatibility analyzers](../native-aot/index.md#aot-compatibility-analyzers).

To generate trim warnings without marking the project as trim-compatible, use `<EnableTrimAnalyzer>true</EnableTrimAnalyzer>` rather than `<IsTrimmable>true</IsTrimmable>`.

### Show all warnings with test app

To show all analysis warnings for a library, the trimmer must analyze the implementation:

* Of the library.
* All dependencies the library uses.

When building and publishing a library:

* The implementations of the dependencies aren't available.
* The available reference assemblies don't have enough information for the trimmer to determine if they're compatible with trimming.

> Because of the dependency limitations, a self-contained test app which uses the library and its dependencies must be created. The test app executable includes all the information the trimmer requires to issue warning on trim incompatibilities in:

Reviewer: ```suggestion
Because of the dependency limitations, a self-contained test app which uses the library and its dependencies must be created. The test app includes all the information the trimmer requires to issue warning on trim incompatibilities in:
```

(The executable part is misleading and technically wrong as well)

* The library code.
* The code that the library references from its dependencies.

***Note:*** If the library has different behavior depending on the target framework, create a trimming test app for each of the target frameworks. For example, if the library uses [conditional compilation](/dotnet/csharp/language-reference/preprocessor-directives#conditional-compilation) such as `#if NET7_0` to change behavior.

> 

Reviewer: This needs to take into account older TFMs where trimming is not possible. Lot of libraries will target `net472`, `netstandard2.0` and then for example `net6.0`. And they will often have ifdefs around these TFMs in the code. So as a library author I would read this as "I should do this" - while in reality, it will not work for anything but 6+.

To create the trimming test app:

* Create a separate console application project.
* Add a reference to the library.
* Modify the project similar to the project shown below using the following list:

### [.NET 6](#tab/net6)

> * Set the `TrimmerDefaultAction` property to `link` with `<TrimmerDefaultAction>link</TrimmerDefaultAction>` n a `<PropertyGroup>` element.  <!-- only diff with .NET7+ -->

Reviewer: ```suggestion
* Set the `TrimmerDefaultAction` property to `link` with `<TrimmerDefaultAction>link</TrimmerDefaultAction>` in a `<PropertyGroup>` element.  <!-- only diff with .NET7+ -->
```

* Add a reference to the library project with `<ProjectReference Include="/Path/To/YourLibrary.csproj" />`.
* Specify the library as a trimmer root assembly with `<TrimmerRootAssembly Include="YourLibraryName" />`.
  * `TrimmerRootAssembly` ensures that every part of the library is analyzed. It tells the trimmer that this assembly is a "root,  ". A "root," assembly means the trimmer analyzes every call in the library, and traverses all code paths that originate from   that assembly. `TrimmerRootAssembly` is necessary in case the library has `[AssemblyMetadata("IsTrimmable", "True")]`.  A   project using `[AssemblyMetadata("IsTrimmable", "True")]` without  `TrimmerRootAssembly` removes the unused library without analyzing it.

### .csproj file

:::code language="xml" source="~/docs/core/deploying/trimming/snippets/MyTestLib6app/MyTestLib6app.csproj":::

### [.NET 7+](#tab/net7plus)

* Add `<PublishTrimmed>true</PublishTrimmed>`.
* Add a reference to the library project with `<ProjectReference Include="/Path/To/YourLibrary.csproj" />`.
* Specify the library as a trimmer root assembly with `<TrimmerRootAssembly Include="YourLibraryName" />`.

>   * `TrimmerRootAssembly` ensures that every part of the library is analyzed. It tells the trimmer that this assembly is a "root". A "root" assembly means the trimmer analyzes every call in the library, and traverses all code paths that originate from that assembly. `TrimmerRootAssembly` is necessary in case the library has `[AssemblyMetadata("IsTrimmable", "True")]`.  A project using `[AssemblyMetadata("IsTrimmable", "True")]` without `TrimmerRootAssembly` removes the unused library without analyzing it.

Reviewer: Same comment as above for the attribute parts - just drop those.

### .csproj file

> :::code language="xml" source="~/docs/core/deploying/trimming/snippets/MyLibrary/MyLibrary.csproj.xml" id="snippet_full":::

Reviewer: I think this is the wrong link - it points to the library project, not the test app project.

To get the most up-to-date trimming warnings and analyzer coverage:

* Install and target the .NET 8 SDK or later.
* When targeting .NET 6 or NET 7, install the .NET 8 SDK or later.

## Enable library trim warnings

Trim warnings in a library can be found with either of the following methods:

* Enabling project-specific trimming using the `IsTrimmable` property.
* Creating a trimming test app that uses the library and enabling trimming for the test app. It's not necessary to reference all the APIs in the library.

We recommend using both approaches. Project-specific trimming is convenient and shows trim warnings for one project, but relies on the references being marked trim-compatible to see all warnings. Trimming a test app is more work, but shows all warnings.

### Enable project-specific trimming

### [.NET 6](#tab/net6)

To get the latest version of the analyzer with the most coverage, install the [.NET 8 SDK](https://dotnet.microsoft.com/download/dotnet) or later. Installing the .NET 8 SDK or later:

* Updates the tooling used to build an app or library and enable trim warnings.
* Provides significant trimming and analyzer improvements over previous .NET SDKs.

> * Does ***not*** require targeting the .NET 8 runtime. Libary authors can continue to target the .NET 6 runtime.

Reviewer: ```suggestion
* Does ***not*** require targeting the .NET 8 runtime. Library authors can continue to target the .NET 6 runtime.
```

* Add a reference to the library project with `<ProjectReference Include="/Path/To/YourLibrary.csproj" />`.
* Specify the library as a trimmer root assembly with `<TrimmerRootAssembly Include="YourLibraryName" />`.
  * `TrimmerRootAssembly` ensures that every part of the library is analyzed. It tells the trimmer that this assembly is a "root". A "root" assembly means the trimmer analyzes every call in the library and traverses all code paths that originate from that assembly.

### .csproj file

:::code language="xml" source="~/docs/core/deploying/trimming/snippets/MyTestLib6app/XMLFile1.xml":::

### [.NET 7](#tab/net7)

* Add `<PublishTrimmed>true</PublishTrimmed>`.
* Add a reference to the library project with `<ProjectReference Include="/Path/To/YourLibrary.csproj" />`.
* Specify the library as a trimmer root assembly with `<TrimmerRootAssembly Include="YourLibraryName" />`.
  * `TrimmerRootAssembly` ensures that every part of the library is analyzed. It tells the trimmer that this assembly is a "root". A "root" assembly means the trimmer analyzes every call in the library and traverses all code paths that originate from that assembly.

### .csproj file

> :::code language="xml" source="~/docs/core/deploying/trimming/snippets/ConsoleApp1/ConsoleApp1.csproj":::

Reviewer: In the .NET 7 tab the code snippets shows a project file targeting `net8.0` TFM. Not a big deal, but it's a bit confusing.

Reviewer: Sorry - noticed the comment below - please ignore.

  * Tells the trimmer that any code that isn't part of a "root" can be trimmed if it's unused. Without this   option:
    * Warnings are issued originating from ***any*** part of a dependency that doesn't set `[AssemblyMetadata  ("IsTrimmable", "Tue")]`
    * The preceding warnings can be issued for code that is unused by the library.

### [.NET 8+](#tab/net8plus)

<!-- exact duplicate of tab7 but forced by builder to have identical tabs -->
* Add `<PublishTrimmed>true</PublishTrimmed>`.
* Add a reference to the library project with `<ProjectReference Include="/Path/To/YourLibrary.csproj" />`.
* Specify the library as a trimmer root assembly with `<TrimmerRootAssembly Include="YourLibraryName" />`.
  * `TrimmerRootAssembly` ensures that every part of the library is analyzed. It tells the trimmer that this assembly is a "root". A "root" assembly means the trimmer analyzes every call in the library and traverses all code paths that originate from that assembly.

### .csproj file

:::code language="xml" source="~/docs/core/deploying/trimming/snippets/ConsoleApp1/ConsoleApp1.csproj":::

**Notes:**

> * [`<TrimMode>full</TrimMode>`](/dotnet/core/compatibility/deployment/7.0/trim-all-assemblies#recommended-action) in the `<PropertyGroup>` tag:

Reviewer: Personally I find this confusing - the project file doesn't have that property (and we do note that here), but initially I was looking for it (since I didn't get to the note that it's not there yet). After reading it all I was like "Why do I care?" - it's the default, I want all of this, so why are you telling me this.

The final bullet point about the "Tells the trimmer" is really confusing, because it's a negation - the subpoints are NOT true in the test sample app, but to know that one has to fully read and understand the sentence before them... I honestly had to read this like 3 times to understand what's going on.

I think this description is a good one to have to document the `TrimMode` property itself, but I would not include it in the description of the test app.

In the following example, an unknown [Type](/dotnet/api/system.type) flows into the annotated method parameter. The unknown `Type` is from a field:

:::code language="csharp" source="~/docs/core/deploying/trimming/snippets/MyLibrary/Class1.cs" id="snippet_UMH":::

Similarly, here the problem is that the field `type` is passed into a parameter with these requirements. It's fixed by adding  [`[DynamicallyAccessedMembers]`](xref:System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute) to the field. `[DynamicallyAccessedMembers]` warns about code that assigns incompatible values to the field. Sometimes this process continues until a public API is annotated, and other times it ends when a concrete type flows into a location with these requirements. For example:

:::code language="csharp" source="~/docs/core/deploying/trimming/snippets/MyLibrary/Class3.cs" id="snippet_UMH2" highlight="1":::

In this case, the trim analysis keeps public methods of <xref:System.Tuple>, and produces further warnings.

## Recommendations

* ***Avoid*** reflection when possible. When using reflection, minimize reflection scope so that it's reachable only from a small part of the library.

> * Annotate code with `DynamicallyAccessedMembers` to statically express the trimming requirements statically when possible.

Reviewer: ```suggestion
* Annotate code with `DynamicallyAccessedMembers` to statically express the trimming requirements when possible.
```

* Consider reorganizing code to make it follow an analyzable pattern that can be annotated with `DynamicallyAccessedMembers`
* When code is incompatible with trimming, annotate it with `RequiresUnreferencedCode` and propagate this annotation to callers until the relevant public APIs are annotated.

> * Avoid using code that uses reflection in a way not understood by the static analysis. For example, reflection in static constructors should be avoided. Using reflection in static constructors result in the warning propagating to all members of the class.

Reviewer: ```suggestion
* Avoid using code that uses reflection in a way not understood by the static analysis. For example, reflection in static constructors should be avoided. Using statically unanalyzable reflection in static constructors result in the warning propagating to all members of the class.
```

It's OK to use reflection which the analysis can figure out - even in a static constructor (in fact there are many use cases for this and we do that in several places in our libraries as well), the problem is trim incompatible code in static constructor - which is hard to reason about who to blame.

* Avoid annotating virtual methods or interface methods. Annotating virtual or interface methods requires all overrides to have matching annotations.
* If an API is mostly trim-incompatible, alternative coding approaches to the API may need to be considered. A common example is reflection-based serializers. In these cases, consider adopting other technology like source generators to produce code that is more easily statically analyzed. For example, see [How to use source generation in System.Text.Json](/dotnet/standard/serialization/system-text-json/source-generation)

<!--

> * When using reflection, minimize reflection scope so that it's reachable only from a small part of the library.

Reviewer: I assume this is to be removed before merging... or is there some usefulness in the commented out text (I'm no expert on the docs formatting rules and such).

Author: Removed

The .NET SDK makes it possible to reduce the size of self-contained apps by [trimming](trim-self-contained.md), which removes unused code from the app and its dependencies. Not all code is compatible with trimming, so .NET 6 provides trim analysis warnings to detect patterns that may break trimmed apps. To resolve warnings originating from the app code, see [resolving trim warnings](#resolve-trim-warnings). This article describes how to prepare libraries for trimming with the aid of these warnings, including recommendations for fixing some common cases.

## Prerequisites

### [.NET 6](#tab/net6)

[.NET 6 SDK](https://dotnet.microsoft.com/download/dotnet)

### [.NET 7](#tab/net7)

[.NET 7 SDK](https://dotnet.microsoft.com/download/dotnet)

### [.NET 8+](#tab/net8plus)

> [.NET 8 SDK](https://dotnet.microsoft.com/download/dotnet)

Author: > Maybe it would be better to include just one download link and remove the 6/7/8 sections? Currently it reads to me like "For .NET 6, the prerequisite is .NET 6", etc.

We've got to get rid of the awful bling for .NET 6. 

<img width="531" alt="image" src="https://github.com/dotnet/docs/assets/3605364/3934a91e-95d7-48b6-b8f2-9bd81d2ba505">

Soon .NET 8 will be released for sure we don't want to wade through the awful mess.

BTW, I'm just getting started on this draft.

All the pivot mess could be avoided if versions were supported like they are in the ASP.NET Core docs.

---

## Enable library trim warnings

There are two ways to find trim warnings in a library:

* Enable project-specific trimming using the `IsTrimmable` property.
* Add the library as a reference to a sample app, and trim the sample app.

Consider doing both. Project-specific trimming is convenient and shows trim warnings for one project, but relies on the references being marked trim-compatible to see all warnings. Trimming a sample app is more work, but will always show all warnings.

### Enable project-specific trimming

### [.NET 6](#tab/net6)

To get the latest version of the analyzer with the most coverage, install the [.NET 7 SDK](https://dotnet.microsoft.com/download/dotnet) or later. Installing the .NET 7 SDK or later:  <!-- Can I replace .NET 7 SDK with .NET 8 SDK before .NET 8 releases? -->

* Updates the tooling used to build an app or library and enable trim warnings.
* Doesn't require targeting the .NET 7 or later runtime.

Set `<IsTrimmable>true</IsTrimmable>` in the project file.

### [.NET 7](#tab/net7)

Set `<IsTrimmable>true</IsTrimmable>` in the project file.

### [.NET 8+](#tab/net8plus)

> Set `<IsTrimmable>true</IsTrimmable>` in the project file.

Author: I need to duplicate something to get rid of the awful .NET 6 note.

> Set `<IsTrimmable>true</IsTrimmable>` in the project file.

Author: fixed

---

Setting `<IsTrimmable>true</IsTrimmable>` marks the assembly as "trimmable" and enables trim warnings. "trimmable" means the project:

* Is considered compatible with trimming.

> * Shouldn't trim warnings when building. When used in a trimmed app, the assembly has its unused members trimmed in the final output.

Reviewer: ```suggestion
* Shouldn't generate trimming related warnings when building. When used in a trimmed app, the assembly has its unused members trimmed in the final output.
```

Or something like that  - the original text is wrong - there's no such thing as "to trim warnings".
